File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/express/add_new_card/OfflineMethodsFragment.java
package com.mercadopago.android.px.internal.features.express.add_new_card;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.SpannableStringBuilder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.TextView;
import com.mercadopago.android.px.R;
import com.mercadopago.android.px.core.BackHandler;
import com.mercadopago.android.px.internal.base.BaseFragment;
import com.mercadopago.android.px.internal.di.Session;
import com.mercadopago.android.px.internal.features.pay_button.PayButton;
import com.mercadopago.android.px.internal.features.pay_button.PayButtonFragment;
import com.mercadopago.android.px.internal.font.PxFont;
import com.mercadopago.android.px.internal.util.TextUtil;
import com.mercadopago.android.px.internal.util.ViewUtils;
import com.mercadopago.android.px.internal.viewmodel.AmountLocalized;
import com.mercadopago.android.px.model.OfflinePaymentTypesMetadata;

public class OfflineMethodsFragment extends BaseFragment<OfflineMethodsPresenter, OfflinePaymentTypesMetadata>
    implements OfflineMethods.OffMethodsView, BackHandler {

    private static final String TAG_EXPLODING_FRAGMENT = "TAG_EXPLODING_FRAGMENT";

    @Nullable /* default */ Animation fadeInAnimation;
    @Nullable /* default */ Animation fadeOutAnimation;
    /* default */ View panIndicator;
    private PayButtonFragment payButtonFragment;
    private TextView totalAmountTextView;
    private View header;

    @NonNull
    public static OfflineMethodsFragment getInstance(@NonNull final OfflinePaymentTypesMetadata model) {
        final OfflineMethodsFragment instance = new OfflineMethodsFragment();
        instance.storeModel(model);
        return instance;
    }

    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater, @Nullable final ViewGroup container,
        @Nullable final Bundle savedInstanceState) {
        return inflater.inflate(R.layout.px_fragment_offline_methods, container, false);
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        header = view.findViewById(R.id.header);
        panIndicator = view.findViewById(R.id.pan_indicator);
        ViewUtils.loadOrHide(View.GONE,
            model.getDisplayInfo() != null ? model.getDisplayInfo().getBottomDescription() : null,
            view.findViewById(R.id.bottom_description));
        payButtonFragment = (PayButtonFragment) getChildFragmentManager().findFragmentById(R.id.pay_button);
        if(payButtonFragment != null) {
            payButtonFragment.disable();
        }
        totalAmountTextView = view.findViewById(R.id.total_amount);
        final ImageView closeImage = view.findViewById(R.id.close);
        closeImage.setOnClickListener(v -> {
            final Activity activity = getActivity();
            if (activity != null) {
                activity.onBackPressed();
            }
        });

        configureRecycler(view.findViewById(R.id.methods));

        if (savedInstanceState == null) {
            presenter.trackOfflineMethodsView(model);
        }

        presenter.updateModel();
    }

    private void configureRecycler(@NonNull final RecyclerView recycler) {
        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
        recycler.setLayoutManager(linearLayoutManager);
        recycler.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull final RecyclerView recyclerView, final int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                final boolean atTop = !recyclerView.canScrollVertically(-1);
                switch (newState) {
                case RecyclerView.SCROLL_STATE_DRAGGING:
                    if (atTop) {
                        panIndicator.clearAnimation();
                        panIndicator.startAnimation(fadeOutAnimation);
                    }
                    break;
                case RecyclerView.SCROLL_STATE_IDLE:
                    if (atTop) {
                        panIndicator.clearAnimation();
                        panIndicator.startAnimation(fadeInAnimation);
                    }
                    break;
                default:
                }
            }

            @Override
            public void onScrolled(@NonNull final RecyclerView recyclerView, final int dx, final int dy) {
                super.onScrolled(recyclerView, dx, dy);
            }
        });
        final DividerItemDecoration decoration =
            new DividerItemDecoration(getContext(), linearLayoutManager.getOrientation());
        decoration.setDrawable(ContextCompat.getDrawable(getContext(), R.drawable.px_item_decorator_divider));
        recycler.addItemDecoration(decoration);

        final OnMethodSelectedListener onMethodSelectedListener = selectedItem -> {
            presenter.selectMethod(selectedItem);
            payButtonFragment.enable();
        };

        final OfflineMethodsAdapter offlineMethodsAdapter =
            new OfflineMethodsAdapter(new FromOfflinePaymentTypesMetadataToOfflineItems(getContext()).map(model),
                onMethodSelectedListener);
        recycler.setAdapter(offlineMethodsAdapter);
    }

    @Override
    public void updateTotalView(@NonNull final AmountLocalized amountLocalized) {
        final Editable editable = new SpannableStringBuilder();
        final Editable editableDescription = new SpannableStringBuilder();
        final String totalText = getString(R.string.px_review_summary_total);
        editable.append(totalText);
        editable.append(TextUtil.SPACE);
        editable.append(amountLocalized.get(getContext()));

        ViewUtils.setFontInSpannable(getContext(), PxFont.SEMI_BOLD, editable);

        totalAmountTextView.setText(editable);

        editableDescription
            .append(totalText)
            .append(TextUtil.SPACE)
            .append(String.valueOf(amountLocalized.getAmount().floatValue()))
            .append(getString(R.string.px_money));

        totalAmountTextView.setContentDescription(editableDescription);
    }

    @Override
    protected OfflineMethodsPresenter createPresenter() {
        final Session session = Session.getInstance();
        return new OfflineMethodsPresenter(session.getPaymentRepository(),
            session.getConfigurationModule().getPaymentSettings(),
            session.getAmountRepository(),
            session.getDiscountRepository(),
            session.getConfigurationModule().getProductIdProvider(),
            model.getPaymentTypes().isEmpty() ? TextUtil.EMPTY : model.getPaymentTypes().get(0).getId(),
            session.getInitRepository(), session.getCongratsRepository());
    }

    @Override
    public boolean handleBack() {
        final boolean isExploding = payButtonFragment.isExploding();
        if (!isExploding) {
            presenter.onBack();
        }
        return isExploding;
    }

    interface OnMethodSelectedListener {
        void onItemSelected(@NonNull final OfflineMethodItem selectedMethod);
    }

    @Override
    public Animation onCreateAnimation(final int transit, final boolean enter, final int nextAnim) {
        final int offset = getResources().getInteger(R.integer.px_long_animation_time);
        final int duration = getResources().getInteger(R.integer.px_shorter_animation_time);
        final Animation animation =
            AnimationUtils.loadAnimation(getContext(), enter ? R.anim.px_fade_in : R.anim.px_fade_out);
        animation.setDuration(duration);
        if (enter) {
            animation.setStartOffset(offset);
        }
        header.startAnimation(animation);
        return super.onCreateAnimation(transit, enter, nextAnim);
    }

    @Override
    public void onSlideSheet(final float offset) {
        header.setAlpha(offset >= 0 ? offset : 0);
    }

    @Override
    public void startKnowYourCustomerFlow(@NonNull final String flowLink) {
        final Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(flowLink));
        startActivity(intent);
    }

    @Override
    public void prePayment(@NonNull PayButton.OnReadyForPaymentCallback callback) {
        presenter.handlePrePaymentAction(callback);
    }

    @Override
    public void onAttach(final Context context) {
        super.onAttach(context);
        final int duration = getResources().getInteger(R.integer.px_shorter_animation_time);
        fadeInAnimation = AnimationUtils.loadAnimation(context, R.anim.px_fade_in);
        fadeInAnimation.setDuration(duration);
        fadeOutAnimation = AnimationUtils.loadAnimation(context, R.anim.px_fade_out);
        fadeOutAnimation.setDuration(duration);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        fadeInAnimation = null;
        fadeOutAnimation = null;
    }
}

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/express/offline_methods/OfflineMethodsFragment.kt
package com.mercadopago.android.px.internal.features.express.offline_methods

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Editable
import android.text.SpannableStringBuilder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.mercadopago.android.px.R
import com.mercadopago.android.px.core.BackHandler
import com.mercadopago.android.px.internal.di.Session
import com.mercadopago.android.px.internal.extensions.addOnLaidOutListener
import com.mercadopago.android.px.internal.extensions.invisible
import com.mercadopago.android.px.internal.extensions.setHeight
import com.mercadopago.android.px.internal.extensions.visible
import com.mercadopago.android.px.internal.features.pay_button.PayButton.OnReadyForPaymentCallback
import com.mercadopago.android.px.internal.features.pay_button.PayButtonFragment
import com.mercadopago.android.px.internal.font.PxFont
import com.mercadopago.android.px.internal.util.TextUtil
import com.mercadopago.android.px.internal.util.ViewUtils
import com.mercadopago.android.px.internal.util.nonNullObserve
import com.mercadopago.android.px.internal.util.nonNullObserveOnce
import com.mercadopago.android.px.internal.view.MPTextView
import com.mercadopago.android.px.internal.viewmodel.AmountLocalized
import com.mercadopago.android.px.model.internal.PaymentConfiguration

class OfflineMethodsFragment : Fragment(), OfflineMethods.View, BackHandler {
    private var fadeInAnimation: Animation? = null
    private var fadeOutAnimation: Animation? = null

    private lateinit var panIndicator: View
    private lateinit var payButtonFragment: PayButtonFragment
    private lateinit var totalAmountTextView: TextView
    private lateinit var header: View
    private lateinit var footer: View
    private lateinit var fakeFooter: View
    private lateinit var bottomDescription: MPTextView
    private lateinit var adapter: OfflineMethodsAdapter

    private lateinit var bottomSheet: View
    private lateinit var bottomSheetBehavior: BottomSheetBehavior<View>
    private lateinit var viewModel: OfflineMethodsViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.px_fragment_offline_methods, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = Session.getInstance().viewModelModule.get(this, OfflineMethodsViewModel::class.java)

        with(view) {
            header = findViewById(R.id.header)
            footer = findViewById(R.id.bottom_sheet_footer)
            fakeFooter = findViewById(R.id.fake_footer)
            panIndicator = findViewById(R.id.pan_indicator)
            bottomSheet = findViewById(R.id.offline_methods_bottom_sheet)
            totalAmountTextView = findViewById(R.id.total_amount)
            bottomDescription = findViewById(R.id.bottom_description)
            configureRecycler(findViewById(R.id.methods))
        }

        payButtonFragment = childFragmentManager.findFragmentById(R.id.pay_button) as PayButtonFragment
        payButtonFragment.disable()

        configureBottomSheet(savedInstanceState)

        with(viewModel) {
            onViewLoaded().nonNullObserveOnce(viewLifecycleOwner) { model -> draw(model) }
            deepLinkLiveData.nonNullObserve(viewLifecycleOwner) { startKnowYourCustomerFlow(it) }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(EXTRA_BOTTOM_SHEET_STATE, bottomSheetBehavior.state)
    }

    private fun draw(model: OfflineMethods.Model) {
        ViewUtils.loadOrHide(View.GONE, model.bottomDescription, bottomDescription)
        footer.addOnLaidOutListener { view -> fakeFooter.setHeight(view.height) }
        updateTotalView(model.amountLocalized)
        adapter.setItems(FromOfflinePaymentTypesMetadataToOfflineItems.map(model.offlinePaymentTypes))
    }

    private fun configureBottomSheet(savedInstanceState: Bundle?) {
        bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet)
        bottomSheetBehavior.isHideable = true
        bottomSheetBehavior.setBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
            override fun onStateChanged(view: View, state: Int) {}
            override fun onSlide(view: View, offset: Float) {
                header.alpha = if (offset >= 0) offset else 0f
                footer.alpha = if (offset >= 0) 1f else if (offset <= -0.5) 0f else 1 + offset * 2
                header.alpha.takeIf { it == 0f }?.run { header.invisible() } ?: run { header.visible() }
                footer.alpha.takeIf { it == 0f }?.run { footer.invisible() } ?: run { footer.visible() }
            }
        })

        savedInstanceState?.let {
            it.getInt(EXTRA_BOTTOM_SHEET_STATE).let { state ->
                bottomSheetBehavior.state = state
                if (state == BottomSheetBehavior.STATE_HIDDEN) footer.invisible() else footer.visible()
                if (state == BottomSheetBehavior.STATE_EXPANDED) header.visible() else header.invisible()
            }
        } ?: run { bottomSheetBehavior.state = BottomSheetBehavior.STATE_HIDDEN }
    }

    private fun configureRecycler(recycler: RecyclerView) {
        val linearLayoutManager = LinearLayoutManager(context).also { recycler.layoutManager = it }
        recycler.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                val atTop = !recyclerView.canScrollVertically(TOP)
                when (newState) {
                    RecyclerView.SCROLL_STATE_DRAGGING -> if (atTop) {
                        panIndicator.clearAnimation()
                        panIndicator.startAnimation(fadeOutAnimation)
                    }
                    RecyclerView.SCROLL_STATE_IDLE -> if (atTop) {
                        panIndicator.clearAnimation()
                        panIndicator.startAnimation(fadeInAnimation)
                    }
                }
            }
        })
        context?.let {
            val decoration = DividerItemDecoration(it, linearLayoutManager.orientation)
            decoration.setDrawable(ContextCompat.getDrawable(it, R.drawable.px_item_decorator_divider)!!)
            recycler.addItemDecoration(decoration)
        }
        adapter = OfflineMethodsAdapter(object : OfflineMethods.OnMethodSelectedListener {
            override fun onItemSelected(selectedItem: OfflineMethodItem) {
                viewModel.onMethodSelected(selectedItem)
                payButtonFragment.enable()
            }
        }).also { recycler.adapter = it }
    }

    private fun updateTotalView(amountLocalized: AmountLocalized) {
        val editable: Editable = SpannableStringBuilder()
        val editableDescription: Editable = SpannableStringBuilder()
        val totalText = getString(R.string.px_review_summary_total)
        context?.let {
            editable.append(totalText)
                .append(TextUtil.SPACE)
                .append(amountLocalized[it])
            ViewUtils.setFontInSpannable(it, PxFont.SEMI_BOLD, editable)
            totalAmountTextView.text = editable
        }
        editableDescription
            .append(totalText)
            .append(TextUtil.SPACE)
            //TODO por que estaba en floatValue?
            .append(amountLocalized.amount.toString())
            .append(getString(R.string.px_money))
        totalAmountTextView.contentDescription = editableDescription
    }

    override fun handleBack(): Boolean {
        val isExploding = payButtonFragment.isExploding()
        val isVisible = bottomSheetBehavior.state != BottomSheetBehavior.STATE_HIDDEN
        if (!isExploding && isVisible) {
            bottomSheetBehavior.state = BottomSheetBehavior.STATE_HIDDEN
            viewModel.onBack()
        }
        return isExploding || isVisible
    }

    override fun onCreateAnimation(transit: Int, enter: Boolean, nextAnim: Int): Animation? {
        val offset = resources.getInteger(R.integer.px_long_animation_time)
        val duration = resources.getInteger(R.integer.px_shorter_animation_time)
        val animation = AnimationUtils.loadAnimation(context, if (enter) R.anim.px_fade_in else R.anim.px_fade_out)
        animation.duration = duration.toLong()
        if (enter) {
            animation.startOffset = offset.toLong()
        }
        header.startAnimation(animation)
        return super.onCreateAnimation(transit, enter, nextAnim)
    }

    override fun showExpanded() {
        viewModel.onSheetShowed()
        bottomSheet.post { bottomSheetBehavior.state = BottomSheetBehavior.STATE_EXPANDED }
    }

    override fun showCollapsed() {
        viewModel.onSheetShowed()
        bottomSheet.post { bottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED }
    }

    private fun startKnowYourCustomerFlow(flowLink: String) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse(flowLink)
        startActivity(intent)
    }

    override fun prePayment(callback: OnReadyForPaymentCallback) {
        viewModel.onPrePayment(callback)
    }

    override fun onPaymentExecuted(configuration: PaymentConfiguration) {
        viewModel.onPaymentExecuted(configuration)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        val duration = resources.getInteger(R.integer.px_shorter_animation_time)
        fadeInAnimation = AnimationUtils.loadAnimation(context, R.anim.px_fade_in)
            .apply { this.duration = duration.toLong() }
        fadeOutAnimation = AnimationUtils.loadAnimation(context, R.anim.px_fade_out)
            .apply { this.duration = duration.toLong() }
    }

    override fun onDetach() {
        super.onDetach()
        fadeInAnimation = null
        fadeOutAnimation = null
    }

    companion object {
        private const val EXTRA_BOTTOM_SHEET_STATE = "bottom_sheet_state"
        private const val TOP = -1
    }
}